File: services/assignment-service/app/eta_client.py
from __future__ import annotations
import httpx

from app.models import DriverCandidate, Settings


async def get_eta(
    client: httpx.AsyncClient,
    settings: Settings,
    candidate: DriverCandidate,
    prep_time_minutes: int,
    traffic_factor: float = 0.3,
    order_id: str | None = None,
) -> int:
    """Return estimatedDeliveryMinutes from the ETA service."""
    payload = {
        "distanceKm": round(candidate.distance_m / 1000, 3),
        "prepTimeMinutes": prep_time_minutes,
        "driverAvailability": 1.0,
        "trafficFactor": traffic_factor,
    }
    if order_id:
        payload["orderId"] = order_id

    r = await client.post(
        f"{settings.eta_service_url}/predict-eta",
        json=payload,
        timeout=5.0,
    )
    r.raise_for_status()
    return r.json()["estimatedDeliveryMinutes"]

File: services/order-service/src/main/java/com/fooddelivery/orderservice/eta/EtaClient.java
package com.fooddelivery.orderservice.eta;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestClient;
import org.springframework.http.MediaType;

@Component
public class EtaClient {

    private final RestClient restClient;
    private final String etaServiceUrl;

    public EtaClient(
            RestClient restClient,
            @Value("${eta.service.url:http://localhost:8084}") String etaServiceUrl) {
        this.restClient = restClient;
        this.etaServiceUrl = etaServiceUrl;
    }

    public int predictEta(
            String orderId,
            double distanceKm,
            int prepTimeMinutes,
            double driverAvailability,
            double trafficFactor) {

        EtaRequest request = new EtaRequest(
                distanceKm,
                prepTimeMinutes,
                driverAvailability,
                trafficFactor,
                orderId
        );

        EtaResponse response = restClient.post()
                .uri(etaServiceUrl + "/predict-eta")
                .contentType(MediaType.APPLICATION_JSON)
                .body(request)
                .retrieve()
                .body(EtaResponse.class);

        if (response == null) {
            throw new IllegalStateException("ETA service returned empty response");
        }

        return response.estimated_delivery_minutes();
    }
}